### ISSUE TITLE
Parameter set from structure

### ISSUE THREAD

**Issue by ChayaSt** [ParmEd/ParmEd#665]
When I try creating a ```ParameterSet``` from ```structure```, the urey bradley terms don't copy over correctly. There are only two atom types in the key (1,3 atoms) instead of all three atom types for the angle. This happens because ```structure.urey_bradleys``` only keeps the 1st and 3rd atom of the angle. 

The fix I'm suggesting is keep the middle atom in ```structure.urey_bradleys``` so that you can get the identical ```ParameterSet```  used to parameterize ```structure````

**Comment by swails** [comment]
Ah, this is a relic from before I started using 3 atom types as the key to Urey-Bradleys.  I'll fix this.

**Comment by ChayaSt** [comment]
Actually, I'm getting a different error now. 

```
ParameterError: Missing angle type for <Angle <Atom CA3 [0]; In PRL 0>--<Atom CA5 [4]; In PRL 0>--<Atom H4 [12]; In PRL 0>; type=<AngleType; k=29.000, theteq=124.600>>
```

This is coming from [here](https://github.com/ParmEd/ParmEd/blob/master/parmed/charmm/psf.py#L520-L525). It should be a ```KeyError``` for the Urey Bradly, not the angle. 

I think this happens because in the original parameter set, The Urey bradley terms that don't exist are set to ```k=0.00; req=0.00```. And it seems like these don't get written to the ```structure.urey_bradleys``` list

**Comment by ChayaSt** [comment]
> Ah, this is a relic from before I started using 3 atom types as the key to Urey-Bradleys. I'll fix this.

I'm almost done fixing this so I'll just PR

**Comment by ChayaSt** [comment]
For the other bug, I can write out UreyBradley terms for all angles and set the values to 0.00 if the UreyBradley term doesn't exist in the structure UreyBradley list

**Comment by swails** [comment]
OK.  I don't want `UreyBradley` to start carrying 3 atoms, though, since that will break a lot of stuff.  The better change (and what I'm mostly done with now) is to just have the `from_structure` constructor look at the bond partners of each atom in the Urey-Bradley to find the central atom.

That's what `UreyBradley.__contains__` does when passed a `Bond`.

**Comment by swails** [comment]
This is actually proving to be a much bigger job than I had anticipated...  Let's both submit our PRs and then we can iterate toward the correct solution from there.  I suspect we will both catch things that the other missed.

**Comment by ChayaSt** [comment]
Yes, it's bigger than I anticipated too.... Mine is not fully working yet
either because of this other bug. I'll submit the PR and we'll iterate.
On Apr 7, 2016 9:21 PM, "Jason Swails" <notifications@github.com> wrote:

This is actually proving to be a much bigger job than I had anticipated...
Let's both submit our PRs and then we can iterate toward the correct
solution from there. I suspect we will both catch things that the other
missed.

—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub
<https://github.com/ParmEd/ParmEd/issues/665#issuecomment-207160496>
